### PR TITLE
Adding New Federation Keys

### PR DESCRIPTION
**Pull request by CodeBrew28**
This mocks the test where we add a new federated key, and the federated executor knows about it, but the gateway doesn't. As long as the field is marked optional, it will be autofilled as an empty argument until it has the correct data. 

Similarly, if you make your federated object identical to the original object, and you add a new field, if the new field is marked optional, it will only send that field over once both servers know about it. If the root server knows about it first, the federated server won't ask for the field until it knows about it. If the federated server knows about it first, it will be autofilled as a nil value until the federated server knows about it.

### CODE REVIEW

File: federation/schema_syncer_test.go [hunk 1]
```diff
@@ -243,3 +243,113 @@ func TestExecutorQueriesWithCustomSchemaSyncer(t *testing.T) {
...
+
+	// Run a federated query and ensure that it works
+	runAndValidateQueryResults(t, ctx, e, query, expectedOutput)
+
+	type UserWithContactInfoNew struct {
+		Id    int64
+		OrgId int64
+		Name  string
+		// When the gateway knows about email, it will fill it correctly
+		Email string `graphql:",optional"`
```
Comment (KevinXing) [inline federation/schema_syncer_test.go:329]: huh, could you point me the code where we deal with the optional field?

**Comment by KevinXing** [comment]
Also, we should force people to make new field as optional.

File: federation/schema_syncer_test.go (hunk 1 above)
Comment (CodeBrew28) [reply federation/schema_syncer_test.go:329]: Yup! https://github.com/samsarahq/thunder/blob/master/graphql/schemabuilder/input.go#L160

File: federation/schema_syncer_test.go [hunk 2]
```diff
@@ -243,3 +243,113 @@ func TestExecutorQueriesWithCustomSchemaSyncer(t *testing.T) {
...
+	})
+
+	type UserWithContactInfo struct {
+		Id    int64
+		OrgId int64
+		Name  string
+	}
+
+	s2 := schemabuilder.NewSchemaWithName("schema2")
+	s2.FederatedFieldFunc("User", func(args struct{ Keys []*UserWithContactInfo }) []*UserWithContactInfo {
```
[the remark this answers is not in the document]
Comment (CodeBrew28) [reply federation/schema_syncer_test.go:272]: Yup!

**Comment by CodeBrew28** [comment]
Yes we should force them, I think that logic can be built into the gateway if it can compare objects between the new and old schema? We should look into adding that enforcement

**Pull request merged**